### PR TITLE
warn on must_use use on async fn's

### PR DESCRIPTION
**Pull request by guswynn**
As referenced in #78149

This only works on `async` fn's for now, I can also look into if I can get `Box<dyn Future>` and `impl Future` working at this level (hir)

### CODE REVIEW

**Comment by eholk** [comment]
Looks good to me!

**Comment by Mark-Simulacrum** [comment]
I suspect this might just be an oversight in this PR, but it seems like this should be an `unused_attributes` *lint*, not an unconditional warning (and a custom message is still fine).

If so, then I suspect it need not receive T-lang approval... though that is a little murky these days :)

**Comment by Mark-Simulacrum** [comment]
T-lang discussed this in today's meeting and agreed that this can move ahead. We may eventually want to make the must_use have behavior related to return of the `.await`'ing the resulting future, but that's not happening immediately (if ever), so we can move ahead here in the meantime.

We briefly discussed my concern around unused_attributes and generally felt positive about its use here, I think.

**Comment by guswynn** [comment]
@wesleywiser what review is outstanding?

**Comment by wesleywiser** [comment]
@guswynn I believe this is still waiting on Mark's feedback to be addressed:

> it seems like this should be an unused_attributes lint, not an unconditional warning (and a custom message is still fine).

**Comment by guswynn** [comment]
@wesleywiser sorry about the delay, was on a break, this should now be ready!

**Review by wesleywiser — APPROVED** [review]
Looks great, thanks @guswynn!

**Comment by wesleywiser** [comment]
> curl: (6) Could not resolve host: ci-mirrors.rust-lang.org

Spurious.

@bors retry

File: compiler/rustc_passes/src/check_attr.rs
Comment (JohnTitor) [inline compiler/rustc_passes/src/check_attr.rs:1065]: In general a message shouldn't have `.` at the end:

```suggestion
                              function, not the value within",
```

Comment (wesleywiser) [reply compiler/rustc_passes/src/check_attr.rs:1065]: Thanks @JohnTitor!

File: src/test/ui/lint/unused/unused-async.stderr
Comment (wesleywiser) [inline src/test/ui/lint/unused/unused-async.stderr:1]:
```suggestion
warning: `must_use` attribute on `async` functions applies to the anonymous `Future` returned by the function, not the value within
```

File: src/test/ui/lint/unused/unused-async.stderr
Comment (wesleywiser) [inline src/test/ui/lint/unused/unused-async.stderr:14]:
```suggestion
warning: `must_use` attribute on `async` functions applies to the anonymous `Future` returned by the function, not the value within
```

**Comment by wesleywiser** [comment]
Blessed tests. 

@bors r+

**Comment by guswynn** [comment]
@wesleywiser you are a legend for blessing those tests, sorry this fell off my radar!

**Comment by wesleywiser** [comment]
@guswynn Thanks for working on it!! 🎉

**Pull request merged**
